File: experiment/compare/gabor_cluster_batch_avg_rl/run_training.py
# -*- coding: utf-8 -*-
"""
@Time    : 2019/9/5 下午3:33
@Author  : 比尔丶盖子
@Email   : 914138410@qq.com
"""
from experiment.compare.gabor_cluster_rl.model import Net
import numpy as np
from util.data_util import loader, convert_label
from util.test_util import run_testing_cluster as run_testing
import torch
from torch.nn import CrossEntropyLoss

batch_size = 40
digits = np.array([2, 3, 4])
cluster_layer_weight_density = 0.01
n_neuron_cluster = 10
n_features_cluster_layer = 50000
n_category = len(digits)
synaptic_th = 0.8  # 中间层和输出层之间连接矩阵的突触阈值
epoch = 5
use_gpu = False
# 构建模型
net = Net(n_features_cluster_layer=n_features_cluster_layer,
          n_neuron_cluster=n_neuron_cluster,
          cluster_layer_weight_density=cluster_layer_weight_density,
          n_category=n_category,
          synaptic_th=synaptic_th)
if torch.cuda.is_available() and use_gpu:
    net = net.cuda()
print(net)
# 数据集
train_loader, test_loader = loader(batch_size=batch_size,
                                   shuffle=True,
                                   flatten=False,
                                   one_hot=False,
                                   digits=digits)
# 损失函数
loss_func = CrossEntropyLoss()


def get_lr(b_label, b_cluster_output):
    lr_map = dict()
    for i in range(len(b_label)):
        label = b_label[i]
        label = int(label.numpy())
        cluster_output = b_cluster_output[i]
        if label in lr_map.keys():
            lr_map[label].append(cluster_output)
        else:
            lr_map[label] = [cluster_output]
    for label in lr_map.keys():
        lr_map[label] = torch.mean(torch.stack(lr_map[label], dim=0), dim=0)
    lr_list = []
    for label in b_label:
        label = int(label.numpy())
        lr_list.append(lr_map[label])
    lr = torch.stack(lr_list, dim=0)
    return lr


for e in range(epoch):
    # b开头的变量表示关于batch的变量
    for step, (b_img, b_label) in enumerate(train_loader):
        net.train()
        if torch.cuda.is_available() and use_gpu:
            b_img = b_img.cuda()
            b_label = b_label.cuda()
        b_label = convert_label(b_label, digits)
        batch_size = len(b_label)
        # forward
        b_output, b_cluster_output = net(b_img)  # shape(batch_size,10)
        cluster_weight = net.cluster.weight
        # backward
        b_predict_prob, b_predict = torch.max(b_output, dim=1)
        b_reward = torch.zeros(batch_size)
        b_reward[b_predict == b_label] = 1
        lr = get_lr(b_label, b_cluster_output)
        for i in range(batch_size):
            reward = b_reward[i]  # 奖励
            predict_prob = b_predict_prob[i]  # 预测的概率range(0,1)
            predict = b_predict[i]
            label = b_label[i]
            cluster_output = b_cluster_output[i]
            weight = net.state_dict()["output.weight"]  # shape(10,n_features_cluster__layer)
            modify_weight = weight[predict, :]  # shape(n_features_cluster__layer)
            rand = torch.rand(modify_weight.shape)
            rand = rand.cuda() if torch.cuda.is_available() and use_gpu else rand
            # 权重值越大有越大的概率被改变
            need_modify_weight = (rand < modify_weight).float()
            # 中间层值越大改变的值越大
            potential = torch.mul(cluster_output, need_modify_weight)  # shape(n_features_cluster__layer)
            if reward:
                modify_weight = modify_weight + lr[i] * (reward - predict_prob) * potential
            else:
                modify_weight = modify_weight - lr[i] * predict_prob * potential
            modify_weight[modify_weight < 0] = 0
            weight[predict, :] = modify_weight
        if step % 100 == 0:
            loss, accuracy = run_testing(net, loss_func, test_loader, use_gpu, digits)
            print('Epoch: ', e, '| train loss: %.4f' % loss, '| test accuracy: %.4f' % accuracy)

File: matlab/run_training.m
function run_training(init_para,net)

[train_img,train_label,test_img,test_label] = load_data(init_para.digits,0.1);
train_len = size(train_img,1);
epoch = init_para.epoch;
% step reward prob verify_acc
check_points = zeros(train_len*epoch,4);
verify_step = init_para.verify_step;
for i=1:epoch
    for j=1:train_len
        % img,label
        batch_img = train_img(j,:)';
        batch_label = train_label(j,:)';
        %forward
        %shape(out_features_cpl,batch_size)
        [output_cpl,b_reward,~,b_predict_prob,b_predict]=forward(net,init_para,batch_img,batch_label);
        %backward
        % update the weights on the final layer
        reward = b_reward;
        predict_prob = b_predict_prob;
        predict = b_predict;
        % shape(1,out_features_cpl)
        modify_weight = net.weight_out(predict, :);  % which synapses will be updated
        % shape(1,out_features_cpl)
        need_modify_weight = rand(1,init_para.out_features_cpl)<modify_weight;
        % shape(1,out_features_cpl)
        potential = output_cpl'.* need_modify_weight;
        depress = ~output_cpl'.* (rand(1,init_para.out_features_cpl)<0.01);
        if reward
            modify_weight = modify_weight + init_para.learning_rate*(reward - predict_prob)*(potential-depress);
        else
            modify_weight = modify_weight - init_para.learning_rate*predict_prob*potential;
        end
        modify_weight = max(modify_weight, 0);
        net.weight_out(predict, :) = modify_weight;
        net.weight_filter_out(predict, :) = net.weight_out(predict, :)>init_para.synaptic_th;
        check_points(j,:) = [j,reward,predict_prob,0];
        if rem(j,verify_step)==0
            start_index = j-verify_step+1;
            end_index = j;
            acc = run_testing(init_para,net,test_img,test_label,-1);
            % acc = mean(check_points(start_index:end_index,2));
            prob = mean(check_points(start_index:end_index,3));
            sprintf("epoch:%d | step:%d | acc:%.4f | prob:%.4f",i,j,acc,prob)
        end
    end
    save("check_points","check_points")
end